src/libopensc/card.c

```diff
@@ -397,41 +397,25 @@ int sc_lock(sc_card_t *card)
 		return r;
 	if (card->lock_count == 0) {
 		if (card->reader->ops->lock != NULL) {
-			r = card->reader->ops->lock(card->reader);
-			while (r == SC_ERROR_CARD_RESET || r == SC_ERROR_READER_REATTACHED) {
+			r = card->reader->ops->lock(card->reader, 0);
+			if (r == SC_ERROR_CARD_RESET) {
 				/* invalidate cache */
 				memset(&card->cache, 0, sizeof(card->cache));
 				card->cache.valid = 0;
-				if (was_reset++ > 4) /* TODO retry a few times */
-					break;
-				r = card->reader->ops->lock(card->reader);
 			}
-			if (r == 0)
-				reader_lock_obtained = 1;
 		}
 		if (r == 0)
 			card->cache.valid = 1;
 	}
 	if (r == 0)
 		card->lock_count++;
 
-	if (r == 0 && was_reset > 0) {
-#ifdef ENABLE_SM
-		if (card->sm_ctx.ops.open)
-			card->sm_ctx.ops.open(card);
-#endif
-	}
-
 	r2 = sc_mutex_unlock(card->ctx, card->mutex);
 	if (r2 != SC_SUCCESS) {
 		sc_log(card->ctx, "unable to release card->mutex lock");
 		r = r != SC_SUCCESS ? r : r2;
 	}
 
-	/* give card driver a chance to do something when reader lock first obtained */
-	if (r == 0 && reader_lock_obtained == 1  && card->ops->card_reader_lock_obtained)
-		r = card->ops->card_reader_lock_obtained(card, was_reset);
```

Comment:
Don't remove that the PIV driver, and any other driver that implements card_reader_lock_obtained
depends on it. For any card that does not  define its on card_reader_lock_obtained the code is a noop.